In [1]:
!pip install split-folders

In [2]:
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import pandas as pd
import random
import cv2
import os
import PIL
import pathlib
import splitfolders

In [3]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras.models import Sequential
from tensorflow.keras.callbacks import ModelCheckpoint, EarlyStopping, ReduceLROnPlateau
from tensorflow.keras.preprocessing.image import ImageDataGenerator


2024-08-11 21:19:27.071940: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-08-11 21:19:27.072092: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-08-11 21:19:27.209136: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered


In [4]:
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.model_selection import StratifiedKFold
from sklearn.preprocessing import LabelEncoder

In [5]:
import wandb
from wandb.integration.keras import WandbCallback
from kaggle_secrets import UserSecretsClient


In [6]:
from kaggle_secrets import UserSecretsClient
user_secrets = UserSecretsClient()
secret_value_0 = user_secrets.get_secret("wandb_api_key")

In [7]:
# Dataset paths
base_ds = '/kaggle/input/potatodataset/Potato Leaf Disease Dataset in Uncontrolled Environment'
base_ds = pathlib.Path(base_ds)

In [8]:
# Prepare the data for Stratified K-Fold
data = []
labels = []

In [9]:
# Gathering file paths and their respective labels
for class_dir in base_ds.glob('*'):
    class_name = class_dir.name
    for img_path in class_dir.glob('*'):
        data.append(str(img_path))
        labels.append(class_name)

In [10]:
# Convert labels to integers
label_encoder = LabelEncoder()
labels = label_encoder.fit_transform(labels)

# Setting up Stratified K-Fold
n_splits = 5  # Number of folds
skf = StratifiedKFold(n_splits=n_splits, shuffle=True, random_state=123)

In [11]:
# Image data generators
train_datagen = ImageDataGenerator(rescale=1./255,
                                   rotation_range=30,
                                   zoom_range=0.15,
                                   width_shift_range=0.2,
                                   height_shift_range=0.2,
                                   shear_range=0.15,
                                   horizontal_flip=True,
                                   fill_mode="nearest")

test_datagen = ImageDataGenerator(rescale=1./255)

In [12]:
!pip install wandb --upgrade

/opt/conda/lib/python3.10/pty.py:89: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
  pid, fd = os.forkpty()


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.1/7.1 MB 43.8 MB/s eta 0:00:0000:0100:01
  Attempting uninstall: wandb
    Found existing installation: wandb 0.17.4
    Uninstalling wandb-0.17.4:
      Successfully uninstalled wandb-0.17.4


In [13]:
import wandb

# Store results from each fold
fold_results = []

# K-Fold Cross-Validation
for fold, (train_idx, val_idx) in enumerate(skf.split(data, labels)):
    print(f"Training fold {fold+1}/{n_splits}...")
    
    # Split data into training and validation sets
    train_files = np.array(data)[train_idx]
    train_labels = np.array(labels)[train_idx]
    val_files = np.array(data)[val_idx]
    val_labels = np.array(labels)[val_idx]
    
    # Convert integer labels back to their string equivalents
    train_labels_str = label_encoder.inverse_transform(train_labels)
    val_labels_str = label_encoder.inverse_transform(val_labels)
    
    # Create training and validation datasets
    train_generator = train_datagen.flow_from_dataframe(
        pd.DataFrame({'filename': train_files, 'class': train_labels_str}),
        x_col='filename',
        y_col='class',
        target_size=(128, 128),
        class_mode='categorical',
        batch_size=16,
        shuffle=True
    )
    
    val_generator = train_datagen.flow_from_dataframe(
        pd.DataFrame({'filename': val_files, 'class': val_labels_str}),
        x_col='filename',
        y_col='class',
        target_size=(128, 128),
        class_mode='categorical',
        batch_size=16,
        shuffle=False
    )
    
    # Model configuration
    CONFIG = dict(
        batch_size=16,
        img_height=128,
        img_width=128,
    )
    
    # Rename the model variable to avoid conflicts
    densenet_model = tf.keras.applications.DenseNet201(weights="imagenet", include_top=False, input_shape=(128, 128, 3))
    densenet_model.trainable = True
    inputs = tf.keras.Input((128, 128, 3))
    x = densenet_model(inputs, training=False)
    x = tf.keras.layers.GlobalAveragePooling2D()(x)
    x = tf.keras.layers.Dense(1024, activation='relu')(x)
    x = tf.keras.layers.Dense(len(label_encoder.classes_), activation='softmax')(x)
    densenet_model = tf.keras.Model(inputs, x)
    
    densenet_model.compile(optimizer=keras.optimizers.Adam(1e-5),
                           loss='categorical_crossentropy', metrics=['accuracy'])
    
    # W&B run
    run = wandb.init(settings=wandb.Settings(start_method="fork"),
                     reinit=True,
                     project='Preet_Potato_kfold',
                     entity="pritmhala03",
                     config=CONFIG,
                     group='DenseNet201',
                     job_type='train')
    
    wandb.config.type = f'fold_{fold+1}'
    
    # Training with manual logging
    for epoch in range(20):
        history = densenet_model.fit(
            train_generator,
            epochs=1,
            validation_data=val_generator
        )
        
        # Log metrics manually to W&B
        wandb.log({
            "epoch": epoch + 1,
            "train_loss": history.history['loss'][0],
            "train_accuracy": history.history['accuracy'][0],
            "val_loss": history.history['val_loss'][0],
            "val_accuracy": history.history['val_accuracy'][0]
        })
    
    # Evaluate model on validation set and log final metrics
    val_loss, val_accuracy = densenet_model.evaluate(val_generator)
    print(f"Fold {fold+1} - Loss: {val_loss}, Accuracy: {val_accuracy * 100:.2f}%")
    fold_results.append((val_loss, val_accuracy))
    wandb.log({"final_val_loss": val_loss, "final_val_accuracy": val_accuracy})
    
    # Save the model in TensorFlow SavedModel format
    # densenet_model.save(f'DenseNet201_fold_{fold+1}', save_format='tf')
    
    # Finish W&B run
    run.finish()

# After all folds
for i, (val_loss, val_accuracy) in enumerate(fold_results):
    print(f"Fold {i+1} - Loss: {val_loss}, Accuracy: {val_accuracy * 100:.2f}%")


Training fold 1/5...
Found 2460 validated image filenames belonging to 7 classes.
Found 616 validated image filenames belonging to 7 classes.
74836368/74836368 ━━━━━━━━━━━━━━━━━━━━ 4s 0us/step


wandb: Logging into wandb.ai. (Learn how to deploy a W&B server locally: https://wandb.me/wandb-server)
wandb: You can find your API key in your browser here: https://wandb.ai/authorize
wandb: Paste an API key from your profile and hit enter, or press ctrl+c to quit:

  ········································


wandb: Appending key for api.wandb.ai to your netrc file: /root/.netrc


/opt/conda/lib/python3.10/site-packages/keras/src/trainers/data_adapters/py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()
I0000 00:00:1723411500.660273     117 device_compiler.h:186] Compiled cluster using XLA!  This line is logged at most once for the lifetime of the process.


  1/154 ━━━━━━━━━━━━━━━━━━━━ 11:05:38 261s/step - accuracy: 0.2500 - loss: 1.8099

W0000 00:00:1723411500.908791     117 graph_launch.cc:671] Fallback to op-by-op mode because memset node breaks graph update


154/154 ━━━━━━━━━━━━━━━━━━━━ 0s 1s/step - accuracy: 0.2699 - loss: 1.9041

W0000 00:00:1723411716.433380     116 graph_launch.cc:671] Fallback to op-by-op mode because memset node breaks graph update


154/154 ━━━━━━━━━━━━━━━━━━━━ 508s 2s/step - accuracy: 0.2703 - loss: 1.9029 - val_accuracy: 0.4302 - val_loss: 1.5545
154/154 ━━━━━━━━━━━━━━━━━━━━ 49s 310ms/step - accuracy: 0.5027 - loss: 1.3647 - val_accuracy: 0.5601 - val_loss: 1.2133
154/154 ━━━━━━━━━━━━━━━━━━━━ 48s 301ms/step - accuracy: 0.6131 - loss: 1.0748 - val_accuracy: 0.6347 - val_loss: 1.0384
154/154 ━━━━━━━━━━━━━━━━━━━━ 48s 299ms/step - accuracy: 0.6565 - loss: 0.9441 - val_accuracy: 0.6883 - val_loss: 0.8721
154/154 ━━━━━━━━━━━━━━━━━━━━ 48s 302ms/step - accuracy: 0.7092 - loss: 0.8174 - val_accuracy: 0.7094 - val_loss: 0.8330
154/154 ━━━━━━━━━━━━━━━━━━━━ 48s 302ms/step - accuracy: 0.7410 - loss: 0.7399 - val_accuracy: 0.7094 - val_loss: 0.8041
154/154 ━━━━━━━━━━━━━━━━━━━━ 47s 296ms/step - accuracy: 0.7700 - loss: 0.6397 - val_accuracy: 0.7159 - val_loss: 0.7774
154/154 ━━━━━━━━━━━━━━━━━━━━ 49s 307ms/step - accuracy: 0.7880 - loss: 0.6000 - val_accuracy: 0.7305 - val_loss: 0.7102
154/154 ━━━━━━━━━━━━━━━━━━━━ 49s 308ms/ste

Training fold 2/5...
Found 2461 validated image filenames belonging to 7 classes.
Found 615 validated image filenames belonging to 7 classes.


wandb: Currently logged in as: pritmhala03. Use `wandb login --relogin` to force relogin


/opt/conda/lib/python3.10/site-packages/keras/src/trainers/data_adapters/py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


  1/154 ━━━━━━━━━━━━━━━━━━━━ 9:21:28 220s/step - accuracy: 0.3125 - loss: 1.6224

W0000 00:00:1723412971.691574     116 graph_launch.cc:671] Fallback to op-by-op mode because memset node breaks graph update


154/154 ━━━━━━━━━━━━━━━━━━━━ 0s 1s/step - accuracy: 0.2859 - loss: 1.8041

W0000 00:00:1723413174.887414     118 graph_launch.cc:671] Fallback to op-by-op mode because memset node breaks graph update


154/154 ━━━━━━━━━━━━━━━━━━━━ 452s 2s/step - accuracy: 0.2863 - loss: 1.8032 - val_accuracy: 0.4146 - val_loss: 1.5262
154/154 ━━━━━━━━━━━━━━━━━━━━ 47s 295ms/step - accuracy: 0.5106 - loss: 1.3476 - val_accuracy: 0.5626 - val_loss: 1.2066
154/154 ━━━━━━━━━━━━━━━━━━━━ 48s 304ms/step - accuracy: 0.6087 - loss: 1.0938 - val_accuracy: 0.6341 - val_loss: 1.0316
154/154 ━━━━━━━━━━━━━━━━━━━━ 48s 303ms/step - accuracy: 0.6605 - loss: 0.9221 - val_accuracy: 0.6667 - val_loss: 0.9001
154/154 ━━━━━━━━━━━━━━━━━━━━ 48s 301ms/step - accuracy: 0.7013 - loss: 0.8317 - val_accuracy: 0.7236 - val_loss: 0.8131
154/154 ━━━━━━━━━━━━━━━━━━━━ 48s 302ms/step - accuracy: 0.7410 - loss: 0.7215 - val_accuracy: 0.7301 - val_loss: 0.7594
154/154 ━━━━━━━━━━━━━━━━━━━━ 47s 296ms/step - accuracy: 0.7867 - loss: 0.6309 - val_accuracy: 0.7577 - val_loss: 0.6981
154/154 ━━━━━━━━━━━━━━━━━━━━ 47s 294ms/step - accuracy: 0.7855 - loss: 0.6021 - val_accuracy: 0.7724 - val_loss: 0.6622
154/154 ━━━━━━━━━━━━━━━━━━━━ 48s 299ms/ste

Training fold 3/5...
Found 2461 validated image filenames belonging to 7 classes.
Found 615 validated image filenames belonging to 7 classes.


/opt/conda/lib/python3.10/site-packages/keras/src/trainers/data_adapters/py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


  1/154 ━━━━━━━━━━━━━━━━━━━━ 9:20:12 220s/step - accuracy: 0.0625 - loss: 2.4362

W0000 00:00:1723414381.799298     118 graph_launch.cc:671] Fallback to op-by-op mode because memset node breaks graph update


154/154 ━━━━━━━━━━━━━━━━━━━━ 0s 959ms/step - accuracy: 0.2528 - loss: 1.9071

W0000 00:00:1723414544.783656     116 graph_launch.cc:671] Fallback to op-by-op mode because memset node breaks graph update


154/154 ━━━━━━━━━━━━━━━━━━━━ 400s 1s/step - accuracy: 0.2533 - loss: 1.9059 - val_accuracy: 0.3951 - val_loss: 1.5608
154/154 ━━━━━━━━━━━━━━━━━━━━ 49s 306ms/step - accuracy: 0.4636 - loss: 1.4078 - val_accuracy: 0.5333 - val_loss: 1.2655
154/154 ━━━━━━━━━━━━━━━━━━━━ 48s 300ms/step - accuracy: 0.5817 - loss: 1.1675 - val_accuracy: 0.6439 - val_loss: 1.0453
154/154 ━━━━━━━━━━━━━━━━━━━━ 48s 304ms/step - accuracy: 0.6523 - loss: 0.9778 - val_accuracy: 0.6358 - val_loss: 0.9385
154/154 ━━━━━━━━━━━━━━━━━━━━ 47s 294ms/step - accuracy: 0.6954 - loss: 0.8450 - val_accuracy: 0.6764 - val_loss: 0.8757
154/154 ━━━━━━━━━━━━━━━━━━━━ 48s 300ms/step - accuracy: 0.7356 - loss: 0.7532 - val_accuracy: 0.6943 - val_loss: 0.7962
154/154 ━━━━━━━━━━━━━━━━━━━━ 48s 304ms/step - accuracy: 0.7712 - loss: 0.6367 - val_accuracy: 0.7301 - val_loss: 0.7305
154/154 ━━━━━━━━━━━━━━━━━━━━ 47s 297ms/step - accuracy: 0.7998 - loss: 0.5738 - val_accuracy: 0.7301 - val_loss: 0.7149
154/154 ━━━━━━━━━━━━━━━━━━━━ 48s 300ms/ste

Training fold 4/5...
Found 2461 validated image filenames belonging to 7 classes.
Found 615 validated image filenames belonging to 7 classes.


/opt/conda/lib/python3.10/site-packages/keras/src/trainers/data_adapters/py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


  1/154 ━━━━━━━━━━━━━━━━━━━━ 9:22:54 221s/step - accuracy: 0.0625 - loss: 2.4640

W0000 00:00:1723415745.619504     118 graph_launch.cc:671] Fallback to op-by-op mode because memset node breaks graph update


154/154 ━━━━━━━━━━━━━━━━━━━━ 0s 938ms/step - accuracy: 0.2563 - loss: 1.8651

W0000 00:00:1723415904.653652     118 graph_launch.cc:671] Fallback to op-by-op mode because memset node breaks graph update


154/154 ━━━━━━━━━━━━━━━━━━━━ 396s 1s/step - accuracy: 0.2568 - loss: 1.8640 - val_accuracy: 0.3951 - val_loss: 1.5965
154/154 ━━━━━━━━━━━━━━━━━━━━ 46s 290ms/step - accuracy: 0.4872 - loss: 1.3883 - val_accuracy: 0.5496 - val_loss: 1.2421
154/154 ━━━━━━━━━━━━━━━━━━━━ 45s 285ms/step - accuracy: 0.5895 - loss: 1.1114 - val_accuracy: 0.6130 - val_loss: 1.0199
154/154 ━━━━━━━━━━━━━━━━━━━━ 46s 290ms/step - accuracy: 0.6594 - loss: 0.9475 - val_accuracy: 0.6699 - val_loss: 0.8684
154/154 ━━━━━━━━━━━━━━━━━━━━ 46s 287ms/step - accuracy: 0.7137 - loss: 0.8169 - val_accuracy: 0.7252 - val_loss: 0.7778
154/154 ━━━━━━━━━━━━━━━━━━━━ 45s 284ms/step - accuracy: 0.7508 - loss: 0.7199 - val_accuracy: 0.7431 - val_loss: 0.7115
154/154 ━━━━━━━━━━━━━━━━━━━━ 46s 288ms/step - accuracy: 0.7524 - loss: 0.6664 - val_accuracy: 0.7691 - val_loss: 0.6626
154/154 ━━━━━━━━━━━━━━━━━━━━ 46s 289ms/step - accuracy: 0.8139 - loss: 0.5546 - val_accuracy: 0.7724 - val_loss: 0.6400
154/154 ━━━━━━━━━━━━━━━━━━━━ 47s 291ms/ste

Training fold 5/5...
Found 2461 validated image filenames belonging to 7 classes.
Found 615 validated image filenames belonging to 7 classes.


/opt/conda/lib/python3.10/site-packages/keras/src/trainers/data_adapters/py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


  1/154 ━━━━━━━━━━━━━━━━━━━━ 8:46:34 207s/step - accuracy: 0.1875 - loss: 1.7720

W0000 00:00:1723417045.416241     117 graph_launch.cc:671] Fallback to op-by-op mode because memset node breaks graph update


154/154 ━━━━━━━━━━━━━━━━━━━━ 0s 870ms/step - accuracy: 0.2684 - loss: 1.7870

W0000 00:00:1723417193.854547     116 graph_launch.cc:671] Fallback to op-by-op mode because memset node breaks graph update


154/154 ━━━━━━━━━━━━━━━━━━━━ 372s 1s/step - accuracy: 0.2689 - loss: 1.7863 - val_accuracy: 0.4049 - val_loss: 1.5222
154/154 ━━━━━━━━━━━━━━━━━━━━ 46s 288ms/step - accuracy: 0.4792 - loss: 1.3897 - val_accuracy: 0.5593 - val_loss: 1.2193
154/154 ━━━━━━━━━━━━━━━━━━━━ 46s 290ms/step - accuracy: 0.5979 - loss: 1.1326 - val_accuracy: 0.6195 - val_loss: 1.0286
154/154 ━━━━━━━━━━━━━━━━━━━━ 46s 288ms/step - accuracy: 0.6481 - loss: 0.9697 - val_accuracy: 0.6537 - val_loss: 0.8821
154/154 ━━━━━━━━━━━━━━━━━━━━ 46s 286ms/step - accuracy: 0.7142 - loss: 0.8153 - val_accuracy: 0.6764 - val_loss: 0.8471
154/154 ━━━━━━━━━━━━━━━━━━━━ 46s 287ms/step - accuracy: 0.7483 - loss: 0.7353 - val_accuracy: 0.7398 - val_loss: 0.7552
154/154 ━━━━━━━━━━━━━━━━━━━━ 47s 292ms/step - accuracy: 0.7586 - loss: 0.6738 - val_accuracy: 0.7528 - val_loss: 0.7132
154/154 ━━━━━━━━━━━━━━━━━━━━ 46s 291ms/step - accuracy: 0.7774 - loss: 0.6313 - val_accuracy: 0.7415 - val_loss: 0.6934
154/154 ━━━━━━━━━━━━━━━━━━━━ 45s 284ms/ste

Fold 1 - Loss: 0.5938485264778137, Accuracy: 78.90%
Fold 2 - Loss: 0.5092435479164124, Accuracy: 82.44%
Fold 3 - Loss: 0.5398438572883606, Accuracy: 81.30%
Fold 4 - Loss: 0.5189046263694763, Accuracy: 81.46%
Fold 5 - Loss: 0.5270301103591919, Accuracy: 82.44%


In [14]:
# After all folds
for i, (val_loss, val_accuracy) in enumerate(fold_results):
    print(f"Fold {i+1} - Loss: {val_loss}, Accuracy: {val_accuracy * 100:.2f}%")

Fold 1 - Loss: 0.5938485264778137, Accuracy: 78.90%
Fold 2 - Loss: 0.5092435479164124, Accuracy: 82.44%
Fold 3 - Loss: 0.5398438572883606, Accuracy: 81.30%
Fold 4 - Loss: 0.5189046263694763, Accuracy: 81.46%
Fold 5 - Loss: 0.5270301103591919, Accuracy: 82.44%


Fold 1 - Loss: 0.5938485264778137, Accuracy: 78.90%

Fold 2 - Loss: 0.5092435479164124, Accuracy: 82.44%

Fold 3 - Loss: 0.5398438572883606, Accuracy: 81.30%

Fold 4 - Loss: 0.5189046263694763, Accuracy: 81.46%

Fold 5 - Loss: 0.5270301103591919, Accuracy: 82.44%

In [15]:
# After all folds: Compute average performance
average_val_loss = np.mean([result[0] for result in fold_results])
average_val_accuracy = np.mean([result[1] for result in fold_results])
print(f"Average Validation Loss: {average_val_loss}, Average Validation Accuracy: {average_val_accuracy * 100:.2f}%")

Average Validation Loss: 0.537774133682251, Average Validation Accuracy: 81.31%


Average Validation Loss: 0.537774133682251,

Average Validation Accuracy: 81.31%